pkg/monitor/agent/agent.go

```diff
@@ -128,6 +128,11 @@ func (a *Agent) SendEvent(typ int, event interface{}) error {
 
 	a.notifyAgentEvent(typ, event)
 
+	// do not marshal notifications if there is no active listeners
+	if !a.hasListeners() {
+		return nil
+	}
+
```

Comment:
I assume that we do the check for listeners after `notifyAgentEvent` because we want to ensure we still run some logic? Would be good to clarify why in the code, so that nobody attempts to optimize this with the same thought that it can be moved before `notifyAgentEvent`.

Reply:
Yeah.. to elaborate on the current terminology: Consumers (invoked by `notifyAgentEvent`) consume in-memory data structures, while Listeners (invoked by `sendToListeners` below) expected the data to already be marshaled. The two queues are separate, so we still want to invoke any Consumers even if there are no listeners.

Fabio and I discussed a bit if/how we could move the Listeners over to the Consumer interface and it should be possible, the tricky part is that we want to avoid marshaling the data for every Listener individually (and only marshal once for all listeners, as the current code does).

Reply:
@christarazi probably it'd be good to expand the comment here a bit, to point out that we don't want to waste cycles marshalling if there are no active listeners. I'll add some lines to clarify the differences between consumers and listeners (and why the latter are much more expensive).